observability/micrometer/src/main/java/org/wildfly/extension/micrometer/MicrometerDeploymentProcessor.java

```diff
@@ -25,38 +25,33 @@
 import org.jboss.as.weld.WeldCapability;
 import org.wildfly.extension.micrometer.api.MicrometerCdiExtension;
 import org.wildfly.extension.micrometer.metrics.MetricRegistration;
-import org.wildfly.extension.micrometer.metrics.MicrometerCollector;
 import org.wildfly.subsystem.service.ServiceDependency;
 import org.wildfly.subsystem.service.ServiceInstaller;
 
 class MicrometerDeploymentProcessor implements DeploymentUnitProcessor {
     static final String WELD_CAPABILITY_NAME = "org.wildfly.weld";
 
-    private final MicrometerSubsystemRegistrar.MicrometerDeploymentConfiguration config;
-
-    MicrometerDeploymentProcessor(MicrometerSubsystemRegistrar.MicrometerDeploymentConfiguration config) {
-        this.config = config;
+    MicrometerDeploymentProcessor() {
     }
 
     @Override
     public void deploy(DeploymentPhaseContext deploymentPhaseContext) throws DeploymentUnitProcessingException {
         final DeploymentUnit deploymentUnit = deploymentPhaseContext.getDeploymentUnit();
 
-        ServiceDependency<MicrometerCollector> collector = ServiceDependency.on(MICROMETER_COLLECTOR_RUNTIME_CAPABILITY.getCapabilityServiceName());
-        Supplier<MetricRegistration> factory = () -> collector.get().collectResourceMetrics(
+        ServiceDependency<MicrometerService> serviceDependency = ServiceDependency.on(MICROMETER_SERVICE_SERVICE_NAME);
+        Supplier<MetricRegistration> factory = () -> serviceDependency.get().collectResourceMetrics(
                 deploymentUnit.getAttachment(DeploymentModelUtils.DEPLOYMENT_RESOURCE),
                 deploymentUnit.getAttachment(DeploymentModelUtils.MUTABLE_REGISTRATION_ATTACHMENT),
-                createDeploymentAddressPrefix(deploymentUnit)::append,
-                this.config.getSubsystemFilter());
+                createDeploymentAddressPrefix(deploymentUnit)::append);
         ServiceInstaller.builder(factory)
                 .requires(ServiceDependency.on(DeploymentCompleteServiceProcessor.serviceName(deploymentUnit.getServiceName())))
-                .requires(collector)
+                .requires(serviceDependency)
                 .asActive()
                 .onStop(MetricRegistration::unregister)
                 .build()
                 .install(deploymentPhaseContext);
 
-        registerCdiExtension(deploymentPhaseContext);
+        registerCdiExtension(deploymentPhaseContext, serviceDependency.get());
```

Comment:
There's no happens-before relationship here. ServiceDependency.on by itself doesn't create one, so this method can run before MICROMETER_SERVICE_SERVICE_NAME) is started and get() can return null. Fairly unlikely, but it's possible, and CI is less likely to catch it because for most tests we have a booted server (so the service is started) and then we deploy. A typical real world use deploys the app as part of boot.

I suggest adding a hack like we did in MicroProfileTelemetryDependencyProcessor.deploy.

Note the TODO WFCORE-6491 in MicroProfileTelemetryDependencyProcessor points to the wrong JIRA; it should be WFCORE-6941

Reply:
@bstansberry I've made the changes, I think. Feel free to ask for more if I missed anything.